api/ombpdf/semdb.py

```diff
@@ -1,12 +1,153 @@
-from lxml import etree
+import logging
 
-from document.xml_importer.importer import import_xml_doc
+from document.models import DocNode, FootnoteCitation
+from document.tree import DocCursor
 
-from . import semdom
+from . import semtree
+
+logger = logging.getLogger(__name__)
+
+
+class PDFAwareCursor(DocCursor):
+    """Extension of DocCursor which also tracks the PDF element which created
+    each node."""
+    @property
+    def pdf_node(self):
+        return self.tree.node[self.identifier].get('pdf_node')
+
+    @pdf_node.setter
+    def pdf_node(self, value):
+        self.tree.node[self.identifier]['pdf_node'] = value
 
 
 def to_db(doc, policy):
-    dom = semdom.to_dom(doc)
-    xml = etree.fromstring(dom.toxml())
-    root = import_xml_doc(policy, xml, ignore_preamble=True)
-    return root
+    writer = DatabaseWriter(policy)
+    builder = semtree.SemanticTreeBuilder(doc, writer)
+    builder.build()
+
+    writer.cursor.nested_set_renumber()
+    # Account for a Django bug which prevents us from saving these
+    # directly due to saving the DocNodes in a bulk_insert
+    for footnote_citation in writer.footnote_citations.values():
+        footnote_citation.doc_node = footnote_citation.doc_node
+        footnote_citation.footnote_node = footnote_citation.footnote_node
+        footnote_citation.save()
+
+    return writer.cursor
+
+
+class DatabaseWriter(semtree.Writer):
+    def __init__(self, policy):
+        self.cursor_stack = [PDFAwareCursor.new_tree('policy', policy=policy)]
+        self.footnote_citations = {}
+
+    @property
+    def cursor(self):
+        return self.cursor_stack[-1]
+
+    @property
+    def sec_level(self):
+        return self.cursor.identifier.split('_').count('sec')
```

Comment:
I'm a little concerned that this code might be a bit too "clever"--at least, it took some thinking for me to mentally verify that this would actually work, and not have weird edge cases.  I can't really think of any easy way to make it more readable/less clever though.

Reply:
I can try to expand it out a bit more and label the intermediate steps.